stage2 regression: cImport fails on macOS Multiprocessing.h copyright symbol (not UTF-8 encoded)

This prevents us from removing `@cImport` hacks/workarounds we have in Mach right now for stage2 compiler support https://github.com/hexops/mach/issues/180

### Zig Version

0.10.0-dev.3885+a7661f115 (aarch64-macos)

### Steps to Reproduce

1. Download the macOS `Multiprocessing.h` header. This is used by GLFW and other libraries indirectly during a cImport. Importantly, this header is NOT UTF-8 encoded on macOS due to it being an older header. If the file is re-encoded as UTF-8, it works. If it is not UTF-8 encoded, however, it does not work (this is a regression with stage2.)
  * `wget https://raw.githubusercontent.com/hexops/sdk-macos-12.0/14613b4917c7059dad8f3789f55bb13a2548f83d/root/System/Library/Frameworks/CoreServices.framework/Versions/A/Frameworks/CarbonCore.framework/Versions/A/Headers/Multiprocessing.h`
2. `main.zig`:

```zig
const foo = @cImport(
    @cInclude("Multiprocessing.h")
);

pub fn main() !void {
    _ = foo.MPCreateTask;
}
```



### Expected Behavior

cImport succeeds like it did with stage1

### Actual Behavior

```zig
% zig build-exe -I . main.zig
zig-cache/o/e02ab29f366ec178f1a79052bec8f555/cimport.zig:2803:42: error: expected ';' after declaration
pub const MPCopyrightNotice = "Copyright ? 1995-2020 Apple Computer, Inc.\n";
                                         ^
zig-cache/o/e02ab29f366ec178f1a79052bec8f555/cimport.zig:2803:42: note: invalid byte: '\xa9'
main.zig:1:13: error: C import failed: AnalysisFail
const foo = @cImport(
            ^~~~~~~~
```

With `-fstage1`, it is happy, however.

--- a/src/translate_c.zig
+++ b/src/translate_c.zig
@@ -5957,20 +5957,36 @@ fn zigifyEscapeSequences(ctx: *Context, m: *MacroCtx) ![]const u8 {
     return bytes[0..i];
 }
 
+/// non-ASCII characters (c > 127) are also treated as non-printable by fmtSliceEscapeLower.
+/// If a C string literal or char literal in a macro is not valid UTF-8, we need to escape
+/// non-ASCII characters so that the Zig source we output will itself be UTF-8.
+fn escapeUnprintables(ctx: *Context, m: *MacroCtx) ![]const u8 {
+    const zigified = try zigifyEscapeSequences(ctx, m);
+    if (std.unicode.utf8ValidateSlice(zigified)) return zigified;
+
+    const formatter = std.fmt.fmtSliceEscapeLower(zigified);
+    const encoded_size = @intCast(usize, std.fmt.count("{s}", .{formatter}));
+    var output = try ctx.arena.alloc(u8, encoded_size);
+    return std.fmt.bufPrint(output, "{s}", .{formatter}) catch |err| switch (err) {
+        error.NoSpaceLeft => unreachable,
+        else => |e| return e,
+    };
+}
+
 fn parseCPrimaryExprInner(c: *Context, m: *MacroCtx, scope: *Scope) ParseError!Node {
     const tok = m.next().?;
     const slice = m.slice();
     switch (tok) {
         .CharLiteral => {
             if (slice[0] != '\'' or slice[1] == '\\' or slice.len == 3) {
-                return Tag.char_literal.create(c.arena, try zigifyEscapeSequences(c, m));
+                return Tag.char_literal.create(c.arena, try escapeUnprintables(c, m));
             } else {
                 const str = try std.fmt.allocPrint(c.arena, "0x{s}", .{std.fmt.fmtSliceHexLower(slice[1 .. slice.len - 1])});
                 return Tag.integer_literal.create(c.arena, str);
             }
         },
         .StringLiteral => {
-            return Tag.string_literal.create(c.arena, try zigifyEscapeSequences(c, m));
+            return Tag.string_literal.create(c.arena, try escapeUnprintables(c, m));
         },
         .IntegerLiteral, .FloatLiteral => {
             return parseCNumLit(c, m);

--- a/test/behavior/translate_c_macros.zig
+++ b/test/behavior/translate_c_macros.zig
@@ -5,6 +5,7 @@ const expectEqual = std.testing.expectEqual;
 const expectEqualStrings = std.testing.expectEqualStrings;
 
 const h = @cImport(@cInclude("behavior/translate_c_macros.h"));
+const latin1 = @cImport(@cInclude("behavior/translate_c_macros_not_utf8.h"));
 
 test "casting to void with a macro" {
     h.IGNORE_ME_1(42);
@@ -133,4 +134,15 @@ test "string literal macro with embedded tab character" {
     if (builtin.zig_backend == .stage2_aarch64) return error.SkipZigTest; // TODO
 
     try expectEqualStrings("hello\t", h.EMBEDDED_TAB);
+}
+
+test "string and char literals that are not UTF-8 encoded. Issue #12784" {
+    if (builtin.zig_backend == .stage2_wasm) return error.SkipZigTest; // TODO
+    if (builtin.zig_backend == .stage2_x86_64) return error.SkipZigTest; // TODO
+    if (builtin.zig_backend == .stage2_c) return error.SkipZigTest; // TODO
+    if (builtin.zig_backend == .stage2_arm) return error.SkipZigTest; // TODO
+    if (builtin.zig_backend == .stage2_aarch64) return error.SkipZigTest; // TODO
+
+    try expectEqual(@as(u8, '\xA9'), latin1.UNPRINTABLE_CHAR);
+    try expectEqualStrings("\xA9\xA9\xA9", latin1.UNPRINTABLE_STRING);
 }

--- a/test/behavior/translate_c_macros_not_utf8.h
+++ b/test/behavior/translate_c_macros_not_utf8.h
@@ -1,0 +1,5 @@
+// Note: This file is encoded with ISO/IEC 8859-1 (latin1), not UTF-8.
+// Do not change the encoding
+
+#define UNPRINTABLE_STRING "©©©"
+#define UNPRINTABLE_CHAR '©'

